notation/sign.ts

```diff
@@ -0,0 +1,96 @@
+import * as fs from 'fs';
+import * as os from 'os';
+import * as path from 'path';
+import * as taskLib from 'azure-pipelines-task-lib/task';
+import * as toolLib from 'azure-pipelines-tool-lib/tool';
+
+import { getDownloadInfo, installFromURL } from './lib/install';
+
+import { getArtifactReferences } from './lib/variables';
+import { getConfigHome } from './lib/fs';
+import { getVaultCredentials } from './lib/credentials';
+
+const AZURE_KV_PLUGIN_VERSION_FILE = 'azure_kv_versions.json';
+const AZURE_KV_VERSION_LOCK_FILE = 'azure_kv_version_lock.json';
+
+export async function sign(): Promise<void> {
+    const artifactRefs = getArtifactReferences();
+    const keyid = taskLib.getInput('keyid', true) || '';
+    const cacerts = taskLib.getInput('cacerts', false) || '';
+    const selfSignedCert = taskLib.getBoolInput('selfSigned', false) === true;
+    const signatureFormat = taskLib.getInput('signatureFormat', false) || 'cose';
+    const allowReferrerAPI = taskLib.getBoolInput('allowReferrersAPI', false);
+    const debug = taskLib.getVariable('system.debug')
+
+    await installPlugin();
+
+    let env = { ...process.env, ...await getVaultCredentials() }
+    if (allowReferrerAPI) {
+        env["NOTATION_EXPERIMENTAL"] = "1"
+    }
+
+    for (const artifactRef of artifactRefs) {
+        // run notation sign
+        const code = await taskLib.tool('notation')
+            .arg(['sign', artifactRef,
+                '--plugin', 'azure-kv',
+                '--id', keyid,
+                '--signature-format', signatureFormat])
+            .argIf(allowReferrerAPI, '--allow-referrers-api')
+            .argIf(cacerts, `--plugin-config=ca_certs=${cacerts}`)
+            .argIf(selfSignedCert, '--plugin-config=self_signed=true')
+            .argIf(debug && debug.toLowerCase() === 'true', '--debug')
+            .exec({ env: env });
+
+        if (code !== 0) {
+            throw new Error(`Failed to sign the artifact: ${artifactRef}`);
+        }
+    }
```

Comment:
Essentially, we are signing the artifact references one by one. Once a sign failed, we fail the whole process instead of continue signing the rest of the references. I'm not sure if this is very helpful. IMO, signing one artifact reference should be independent of signing another. Can we finish the whole process and print out two lists in the end: `[references got signed]` and `[references failed to sign]`?

Reply:
Updated.